# Task 1: customer segments and their chip purchasing behaviour

In [10]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import datetime

In [11]:
transactiondata = pd.read_excel(r"\Users\user\Downloads\light_hall\DASL1_Level 5\QVI_transaction_data.xlsx")

In [12]:
customerdata = pd.read_csv(r"\Users\user\Downloads\light_hall\DASL1_Level 5\QVI_purchase_behaviour.csv")

In [13]:
transdata = transactiondata.copy()
transdata

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0       43390          1            1000       1         5   
1       43599          1            1307     348        66   
2       43605          1            1343     383        61   
3       43329          2            2373     974        69   
4       43330          2            2426    1038       108   
...       ...        ...             ...     ...       ...   
264831  43533        272          272319  270088        89   
264832  43325        272          272358  270154        74   
264833  43410        272          272379  270187        51   
264834  43461        272          272379  270188        42   
264835  43365        272          272380  270189        74   

                                       PROD_NAME  PROD_QTY  TOT_SALES  
0         Natural Chip        Compny SeaSalt175g         2        6.0  
1                       CCs Nacho Cheese    175g         3        6.3  
2         Smiths Crinkle Cut  Chips Chicken 170g         2        2.9  
3         Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0  
4       Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8  
...                                          ...       ...        ...  
264831   Kettle Sweet Chilli And Sour Cream 175g         2       10.8  
264832             Tostitos Splash Of  Lime 175g         1        4.4  
264833                  Doritos Mexicana    170g         2        8.8  
264834   Doritos Corn Chip Mexican Jalapeno 150g         2        7.8  
264835             Tostitos Splash Of  Lime 175g         2        8.8  

[264836 rows x 8 columns]

In [14]:
customerdata.head()

LYLTY_CARD_NBR               LIFESTAGE PREMIUM_CUSTOMER
0            1000   YOUNG SINGLES/COUPLES          Premium
1            1002   YOUNG SINGLES/COUPLES       Mainstream
2            1003          YOUNG FAMILIES           Budget
3            1004   OLDER SINGLES/COUPLES       Mainstream
4            1005  MIDAGE SINGLES/COUPLES       Mainstream

In [15]:
transdata.isnull().any()

DATE              False
STORE_NBR         False
LYLTY_CARD_NBR    False
TXN_ID            False
PROD_NBR          False
PROD_NAME         False
PROD_QTY          False
TOT_SALES         False
dtype: bool

In [16]:
customerdata.isnull().any()

LYLTY_CARD_NBR      False
LIFESTAGE           False
PREMIUM_CUSTOMER    False
dtype: bool

In [17]:
customerdata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 72637 entries, 0 to 72636
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   LYLTY_CARD_NBR    72637 non-null  int64 
 1   LIFESTAGE         72637 non-null  object
 2   PREMIUM_CUSTOMER  72637 non-null  object
dtypes: int64(1), object(2)
memory usage: 1.7+ MB


In [18]:
transdata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 264836 entries, 0 to 264835
Data columns (total 8 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   DATE            264836 non-null  int64  
 1   STORE_NBR       264836 non-null  int64  
 2   LYLTY_CARD_NBR  264836 non-null  int64  
 3   TXN_ID          264836 non-null  int64  
 4   PROD_NBR        264836 non-null  int64  
 5   PROD_NAME       264836 non-null  object 
 6   PROD_QTY        264836 non-null  int64  
 7   TOT_SALES       264836 non-null  float64
dtypes: float64(1), int64(6), object(1)
memory usage: 16.2+ MB


we can see the date is of the integer format so we change it

# Change date from xls integer dates to date format in customer data
transdata['DATE'] = pd.to_datetime(transdata['DATE'], unit='D', origin='1899-12-30')

In [19]:
#to examine the PROD NAME colmn

transdata['PROD_NAME'].unique()

array(['Natural Chip        Compny SeaSalt175g',
       'CCs Nacho Cheese    175g',
       'Smiths Crinkle Cut  Chips Chicken 170g',
       'Smiths Chip Thinly  S/Cream&Onion 175g',
       'Kettle Tortilla ChpsHny&Jlpno Chili 150g',
       'Old El Paso Salsa   Dip Tomato Mild 300g',
       'Smiths Crinkle Chips Salt & Vinegar 330g',
       'Grain Waves         Sweet Chilli 210g',
       'Doritos Corn Chip Mexican Jalapeno 150g',
       'Grain Waves Sour    Cream&Chives 210G',
       'Kettle Sensations   Siracha Lime 150g',
       'Twisties Cheese     270g', 'WW Crinkle Cut      Chicken 175g',
       'Thins Chips Light&  Tangy 175g', 'CCs Original 175g',
       'Burger Rings 220g', 'NCC Sour Cream &    Garden Chives 175g',
       'Doritos Corn Chip Southern Chicken 150g',
       'Cheezels Cheese Box 125g', 'Smiths Crinkle      Original 330g',
       'Infzns Crn Crnchers Tangy Gcamole 110g',
       'Kettle Sea Salt     And Vinegar 175g',
       'Smiths Chip Thinly  Cut Original 175g', 'K

In [20]:
# remove digits from the prod_name
# remove & characters which are either toppings or flavors

Prod_name = transdata['PROD_NAME'].str.strip().str.replace(r'[0-9]+[gG]', '').str.replace(r'&', '');

display(Prod_name)

C:\Users\user\AppData\Local\Temp\ipykernel_15632\1182408522.py:4: FutureWarning: The default value of regex will change from True to False in a future version.
  Prod_name = transdata['PROD_NAME'].str.strip().str.replace(r'[0-9]+[gG]', '').str.replace(r'&', '');


0          Natural Chip        Compny SeaSalt
1                        CCs Nacho Cheese    
2          Smiths Crinkle Cut  Chips Chicken 
3           Smiths Chip Thinly  S/CreamOnion 
4         Kettle Tortilla ChpsHnyJlpno Chili 
                         ...                 
264831    Kettle Sweet Chilli And Sour Cream 
264832              Tostitos Splash Of  Lime 
264833                   Doritos Mexicana    
264834    Doritos Corn Chip Mexican Jalapeno 
264835              Tostitos Splash Of  Lime 
Name: PROD_NAME, Length: 264836, dtype: object

In [21]:
# Count the frequencies of words in product names and display counts in descending order
word_counts = pd.Series(' '.join(Prod_name).split()).value_counts()

with pd.option_context('display.max_rows', None): # show all rows
  display(word_counts)

Chips                49770
Kettle               41288
Smiths               28860
Salt                 27976
Cheese               27890
Pringles             25102
Doritos              24962
Crinkle              23960
Corn                 22063
Original             21560
Cut                  20754
Chip                 18645
Chicken              18577
Salsa                18094
Sea                  14145
Thins                14075
Chilli               13895
Sour                 13882
Crisps               12607
Vinegar              12402
RRD                  11894
Infuzions            11057
Supreme              10963
WW                   10320
Popd                  9693
Cobs                  9693
Tortilla              9580
Tostitos              9471
Twisties              9454
Sensations            9429
Dip                   9324
Paso                  9324
El                    9324
Old                   9324
Sweet                 7883
Lime                  7852
Chives                7846
T

In [22]:
# Remove salsas from the dataset
transdata = transdata[transdata['PROD_NAME'].str.contains(r"[Ss]alsa") == False]
transdata.shape # check for a reduction in no of rows 

(246742, 8)

In [23]:
# Now we can create summaries of the data (eg min, max, mean) to see if there are any obvious outliers in the data 
# and if there are any nulls in any of the columns

# #Create summaries of the transaction data
transdata.describe()

DATE      STORE_NBR  LYLTY_CARD_NBR        TXN_ID  \
count  246742.000000  246742.000000    2.467420e+05  2.467420e+05   
mean    43464.054875     135.051098    1.355310e+05  1.351311e+05   
std       105.396691      76.787096    8.071528e+04  7.814772e+04   
min     43282.000000       1.000000    1.000000e+03  1.000000e+00   
25%     43373.000000      70.000000    7.001500e+04  6.756925e+04   
50%     43464.000000     130.000000    1.303670e+05  1.351830e+05   
75%     43555.000000     203.000000    2.030840e+05  2.026538e+05   
max     43646.000000     272.000000    2.373711e+06  2.415841e+06   

            PROD_NBR       PROD_QTY      TOT_SALES  
count  246742.000000  246742.000000  246742.000000  
mean       56.351789       1.908062       7.321322  
std        33.695428       0.659831       3.077828  
min         1.000000       1.000000       1.700000  
25%        26.000000       2.000000       5.800000  
50%        53.000000       2.000000       7.400000  
75%        87.000000       2.000000       8.800000  
max       114.000000     200.000000     650.000000

In [24]:
#using binning to also check for outliers
transdata["PROD_QTY"].value_counts(bins=4).sort_index()

(0.8, 50.75]       246740
(50.75, 100.5]          0
(100.5, 150.25]         0
(150.25, 200.0]         2
Name: PROD_QTY, dtype: int64

In [25]:
# Filter the entries that have 200 packets. 
transdata.loc[transdata['PROD_QTY'] == 200.0]

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
69762  43331        226          226000  226201         4   
69763  43605        226          226000  226210         4   

                              PROD_NAME  PROD_QTY  TOT_SALES  
69762  Dorito Corn Chp     Supreme 380g       200      650.0  
69763  Dorito Corn Chp     Supreme 380g       200      650.0

In [26]:
# The same customer has made these transactions. They could have been for commercial 
# purposes so we can check to see if they made any other purchases.

# Filter the entires by the customer
transdata.loc[transdata['LYLTY_CARD_NBR'] == 226000]

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
69762  43331        226          226000  226201         4   
69763  43605        226          226000  226210         4   

                              PROD_NAME  PROD_QTY  TOT_SALES  
69762  Dorito Corn Chp     Supreme 380g       200      650.0  
69763  Dorito Corn Chp     Supreme 380g       200      650.0

It looks like this is the only purchase they have made so we will remove these transactions from the dataset.

In [27]:
# Remove the transactions
transdata = transdata[transdata['LYLTY_CARD_NBR'] != 226000]
transdata.shape

(246740, 8)

In [28]:
# Recheck the data summary 
transdata.describe()

DATE      STORE_NBR  LYLTY_CARD_NBR        TXN_ID  \
count  246740.000000  246740.000000    2.467400e+05  2.467400e+05   
mean    43464.054843     135.050361    1.355303e+05  1.351304e+05   
std       105.396396      76.786971    8.071520e+04  7.814760e+04   
min     43282.000000       1.000000    1.000000e+03  1.000000e+00   
25%     43373.000000      70.000000    7.001500e+04  6.756875e+04   
50%     43464.000000     130.000000    1.303670e+05  1.351815e+05   
75%     43555.000000     203.000000    2.030832e+05  2.026522e+05   
max     43646.000000     272.000000    2.373711e+06  2.415841e+06   

            PROD_NBR       PROD_QTY      TOT_SALES  
count  246740.000000  246740.000000  246740.000000  
mean       56.352213       1.906456       7.316113  
std        33.695235       0.342499       2.474897  
min         1.000000       1.000000       1.700000  
25%        26.000000       2.000000       5.800000  
50%        53.000000       2.000000       7.400000  
75%        87.000000       2.000000       8.800000  
max       114.000000       5.000000      29.500000

In [29]:
# Count transactions by date to see if there are any missing days 
transdata['DATE'].nunique()

364

**There** **are 365 days in a year but in the DATE column there are only 364 unique values so one** is **missing**

In [30]:
# There is one day of data missing. First check the range of dates by sorting in time order. 
transdata.sort_values(by='DATE')

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
9161    43282         88           88140   86914        25   
155442  43282         60           60276   57330         3   
181349  43282        199          199014  197623       104   
229948  43282         35           35052   31630        11   
104647  43282         72           72104   71038        20   
...       ...        ...             ...     ...       ...   
10254   43646        112          112141  114611        98   
113220  43646        207          207155  205513        99   
229182  43646         10           10140    9882        12   
229015  43646          6            6258    6047        29   
262768  43646        183          183196  185975        22   

                                       PROD_NAME  PROD_QTY  TOT_SALES  
9161              Pringles SourCream  Onion 134g         2        7.4  
155442  Kettle Sensations   Camembert & Fig 150g         2        9.2  
181349  Infuzions Thai SweetChili PotatoMix 110g         2        7.6  
229948                  RRD Pc Sea Salt     165g         1        3.0  
104647          Doritos Cheese      Supreme 330g         2       11.4  
...                                          ...       ...        ...  
10254     NCC Sour Cream &    Garden Chives 175g         2        6.0  
113220          Pringles Sthrn FriedChicken 134g         2        7.4  
229182   Natural Chip Co     Tmato Hrb&Spce 175g         2        6.0  
229015            French Fries Potato Chips 175g         1        3.0  
262768    Thins Chips         Originl saltd 175g         2        6.6  

[246740 rows x 8 columns]

We can see that the dates range from 1 Jul 2018 to 30 Jun 2019. Now we want to check through the year of dates to see which day the data is missing.

In [31]:
# Generate a list of dates with transactions in ascending order 
date_counts = transdata.groupby('DATE').size()

# Then compare to a full list of dates within the same range to find differences between them 
pd.date_range(start = '2018-07-01', end = '2019-06-30' ).difference(date_counts.index)

DatetimeIndex(['2018-07-01', '2018-07-02', '2018-07-03', '2018-07-04',
               '2018-07-05', '2018-07-06', '2018-07-07', '2018-07-08',
               '2018-07-09', '2018-07-10',
               ...
               '2019-06-21', '2019-06-22', '2019-06-23', '2019-06-24',
               '2019-06-25', '2019-06-26', '2019-06-27', '2019-06-28',
               '2019-06-29', '2019-06-30'],
              dtype='datetime64[ns]', length=365, freq='D')

The missing date is Christmas day, a public holiday, so it is expected that there are no sales on this day. Now we move onto creating other features such as the pack size, and checking this for any outliers.

# add a new column to data with packet sizes and extract sizes from product name column

In [32]:
transdata.insert(8, "PACKET_SIZE", transdata['PROD_NAME'].str.extract('(\d+)').astype(float), True)

In [33]:
transdata.head()

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0  43390          1            1000       1         5   
1  43599          1            1307     348        66   
2  43605          1            1343     383        61   
3  43329          2            2373     974        69   
4  43330          2            2426    1038       108   

                                  PROD_NAME  PROD_QTY  TOT_SALES  PACKET_SIZE  
0    Natural Chip        Compny SeaSalt175g         2        6.0        175.0  
1                  CCs Nacho Cheese    175g         3        6.3        175.0  
2    Smiths Crinkle Cut  Chips Chicken 170g         2        2.9        170.0  
3    Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0        175.0  
4  Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8        150.0

In [34]:
# Sort by packet sizes to check for outliers 
transdata.sort_values(by='PACKET_SIZE')

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
40783   43368         97           97067   96696        38   
42461   43590        110          110030  111890        38   
176183  43464         82           82183   81660        38   
227309  43437        236          236091  239098        38   
42418   43409        109          109217  111470        38   
...       ...        ...             ...     ...       ...   
192034  43536        100          100121   99145         4   
255797  43484        235          235098  238018         4   
233814  43489        151          151102  149810         4   
131573  43290        213          213087  212416         4   
102409  43593         43           43184   39874         4   

                                      PROD_NAME  PROD_QTY  TOT_SALES  \
40783   Infuzions Mango     Chutny Papadums 70g         2        4.8   
42461   Infuzions Mango     Chutny Papadums 70g         2        4.8   
176183  Infuzions Mango     Chutny Papadums 70g         2        4.8   
227309  Infuzions Mango     Chutny Papadums 70g         2        4.8   
42418   Infuzions Mango     Chutny Papadums 70g         2        4.8   
...                                         ...       ...        ...   
192034         Dorito Corn Chp     Supreme 380g         2       13.0   
255797         Dorito Corn Chp     Supreme 380g         2       13.0   
233814         Dorito Corn Chp     Supreme 380g         1        6.5   
131573         Dorito Corn Chp     Supreme 380g         2       13.0   
102409         Dorito Corn Chp     Supreme 380g         2       13.0   

        PACKET_SIZE  
40783          70.0  
42461          70.0  
176183         70.0  
227309         70.0  
42418          70.0  
...             ...  
192034        380.0  
255797        380.0  
233814        380.0  
131573        380.0  
102409        380.0  

[246740 rows x 9 columns]

In [35]:
# Add a column to extract the first word of each product name to. 
transdata.insert(9, "BRAND_NAME",transdata['PROD_NAME'].str.split().str.get(0), True)

In [36]:
transdata.head()

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0  43390          1            1000       1         5   
1  43599          1            1307     348        66   
2  43605          1            1343     383        61   
3  43329          2            2373     974        69   
4  43330          2            2426    1038       108   

                                  PROD_NAME  PROD_QTY  TOT_SALES  PACKET_SIZE  \
0    Natural Chip        Compny SeaSalt175g         2        6.0        175.0   
1                  CCs Nacho Cheese    175g         3        6.3        175.0   
2    Smiths Crinkle Cut  Chips Chicken 170g         2        2.9        170.0   
3    Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0        175.0   
4  Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8        150.0   

  BRAND_NAME  
0    Natural  
1        CCs  
2     Smiths  
3     Smiths  
4     Kettle

In [37]:
# Then print all unique entries to check the brand names created
transdata["BRAND_NAME"].unique()

array(['Natural', 'CCs', 'Smiths', 'Kettle', 'Grain', 'Doritos',
       'Twisties', 'WW', 'Thins', 'Burger', 'NCC', 'Cheezels', 'Infzns',
       'Red', 'Pringles', 'Dorito', 'Infuzions', 'Smith', 'GrnWves',
       'Tyrrells', 'Cobs', 'French', 'RRD', 'Tostitos', 'Cheetos',
       'Woolworths', 'Snbts', 'Sunbites'], dtype=object)

Some brand names have been doubled up. Replace all contractions and double ups with their full name.

In [38]:
# Create a function to identify the string replacements needed.
def replace_brandname(line):
    name = line['BRAND_NAME']
    if name == "Infzns":
        return "Infuzions"
    elif name == "Red":
        return "Red Rock Deli"
    elif name == "RRD":
        return "Red Rock Deli"
    elif name == "Grain":
        return "Grain Waves"
    elif name == "GrnWves":
        return "Grain Waves"
    elif name == "Snbts":
        return "Sunbites"
    elif name == "Natural":
        return "Natural Chip Co"
    elif name == "NCC":
        return "Natural Chip Co"
    elif name == "WW":
        return "Woolworths"
    elif name == "Smith":
        return "Smiths"
    elif name == "Dorito":
        return "Doritos"
    else:
        return name 

# Then apply the function to clean the brand names
transdata["BRAND_NAME"] = transdata.apply(lambda line: replace_brandname(line), axis=1)

# Check that there are no duplicate brands 
transdata["BRAND_NAME"].unique()

array(['Natural Chip Co', 'CCs', 'Smiths', 'Kettle', 'Grain Waves',
       'Doritos', 'Twisties', 'Woolworths', 'Thins', 'Burger', 'Cheezels',
       'Infuzions', 'Red Rock Deli', 'Pringles', 'Tyrrells', 'Cobs',
       'French', 'Tostitos', 'Cheetos', 'Sunbites'], dtype=object)

Now we want to examine the customer data. We can generate summaries and check the categories in this dataset.

In [40]:
cust_df = customerdata.copy()
cust_df.head()

LYLTY_CARD_NBR               LIFESTAGE PREMIUM_CUSTOMER
0            1000   YOUNG SINGLES/COUPLES          Premium
1            1002   YOUNG SINGLES/COUPLES       Mainstream
2            1003          YOUNG FAMILIES           Budget
3            1004   OLDER SINGLES/COUPLES       Mainstream
4            1005  MIDAGE SINGLES/COUPLES       Mainstream

In [43]:
# Rename "PREMIUM_CUSTOMER" to "MEMBER_TYPE" for easier identification of the column data
cust_df = cust_df.rename(columns={'PREMIUM_CUSTOMER': 'MEMBER_TYPE'})

In [44]:
# Check the summary of the customer data
cust_df.describe()

LYLTY_CARD_NBR
count    7.263700e+04
mean     1.361859e+05
std      8.989293e+04
min      1.000000e+03
25%      6.620200e+04
50%      1.340400e+05
75%      2.033750e+05
max      2.373711e+06

In [48]:
cust_df.head()

LYLTY_CARD_NBR               LIFESTAGE MEMBER_TYPE
0            1000   YOUNG SINGLES/COUPLES     Premium
1            1002   YOUNG SINGLES/COUPLES  Mainstream
2            1003          YOUNG FAMILIES      Budget
3            1004   OLDER SINGLES/COUPLES  Mainstream
4            1005  MIDAGE SINGLES/COUPLES  Mainstream

In [50]:
# Check the entries in the member type and lifestage columns
cust_df["MEMBER_TYPE"].unique()

array(['Premium', 'Mainstream', 'Budget'], dtype=object)

In [53]:
cust_df["LIFESTAGE"].unique()

array(['YOUNG SINGLES/COUPLES', 'YOUNG FAMILIES', 'OLDER SINGLES/COUPLES',
       'MIDAGE SINGLES/COUPLES', 'NEW FAMILIES', 'OLDER FAMILIES',
       'RETIREES'], dtype=object)

Now that the customer dataset looks fine, we want to add this information to the transactions dataset.

In [54]:
cust_df.shape

(246740, 10)

In [55]:
transdata.shape

(246740, 10)

In [61]:
# Join the customer and transaction datasets, and sort transactons by date 

merged_data = pd.merge(left = cust_df, right = transdata, how = 'right', on = 'LYLTY_CARD_NBR')

In [62]:
merged_data.head()

LYLTY_CARD_NBR               LIFESTAGE MEMBER_TYPE   DATE  STORE_NBR  \
0            1000   YOUNG SINGLES/COUPLES     Premium  43390          1   
1            1307  MIDAGE SINGLES/COUPLES      Budget  43599          1   
2            1343  MIDAGE SINGLES/COUPLES      Budget  43605          1   
3            2373  MIDAGE SINGLES/COUPLES      Budget  43329          2   
4            2426  MIDAGE SINGLES/COUPLES      Budget  43330          2   

   TXN_ID  PROD_NBR                                 PROD_NAME  PROD_QTY  \
0       1         5    Natural Chip        Compny SeaSalt175g         2   
1     348        66                  CCs Nacho Cheese    175g         3   
2     383        61    Smiths Crinkle Cut  Chips Chicken 170g         2   
3     974        69    Smiths Chip Thinly  S/Cream&Onion 175g         5   
4    1038       108  Kettle Tortilla ChpsHny&Jlpno Chili 150g         3   

   TOT_SALES  PACKET_SIZE       BRAND_NAME  
0        6.0        175.0  Natural Chip Co  
1        6.3        175.0              CCs  
2        2.9        170.0           Smiths  
3       15.0        175.0           Smiths  
4       13.8        150.0           Kettle

In [69]:
# Check for nulls in the full dataset
full_df.isnull().values.any()

False

In [70]:
# looks like all the data is reasonable so export to CSV
full_df.to_csv('QVI_fulldata.csv')